File: backend/app/repositories/entities.py
from __future__ import annotations

from typing import Any


class EntitiesRepository:
    def __init__(self, *, supabase: Any):
        self._supa = supabase

    def fetch_video_ids_in_window(
        self,
        *,
        start_iso: str,
        end_iso: str,
        limit: int,
    ) -> list[str]:
        resp = (
            self._supa.table("videos")
            .select("video_id")
            .gte("published_at", start_iso)
            .lte("published_at", end_iso)
            .order("published_at", desc=True)
            .limit(int(limit))
            .execute()
        )
        return [str(r.get("video_id")) for r in (resp.data or []) if isinstance(r, dict) and r.get("video_id")]

    def fetch_summaries_for_video_ids(
        self,
        *,
        video_ids: list[str],
        limit: int,
    ) -> list[dict[str, Any]]:
        if not video_ids:
            return []
        resp = (
            self._supa.table("summaries")
            .select("video_id,ticker,summary")
            .in_("video_id", [str(x) for x in video_ids])
            .limit(int(limit))
            .execute()
        )
        rows = resp.data or []
        return [r for r in rows if isinstance(r, dict)]

    def fetch_recent_videos(
        self,
        *,
        start_iso: str,
        end_iso: str,
        limit: int,
    ) -> list[dict[str, Any]]:
        resp = (
            self._supa.table("videos")
            .select("video_id,published_at,video_url,channel,title")
            .gte("published_at", start_iso)
            .lte("published_at", end_iso)
            .order("published_at", desc=True)
            .limit(int(limit))
            .execute()
        )
        rows = resp.data or []
        return [r for r in rows if isinstance(r, dict)]

    def fetch_entity_summary_rows(
        self,
        *,
        symbol: str,
        allowed_video_ids: list[str],
        limit: int,
    ) -> list[dict[str, Any]]:
        if not allowed_video_ids:
            return []
        resp = (
            self._supa.table("summaries")
            .select(
                "video_id,ticker,summary,created_at,"
                "videos(video_url,video_id,channel,title,published_at)"
            )
            .eq("ticker", symbol)
            .order("created_at", desc=True)
            .limit(int(limit))
            .in_("video_id", [str(x) for x in allowed_video_ids])
            .execute()
        )
        rows = resp.data or []
        return [r for r in rows if isinstance(r, dict)]

File: backend/internal/repo/entities.go
package repo

import (
	"fmt"

	"yunews/backend/internal/supabase"
)

// EntitiesRepository accesses the videos and summaries tables for entity/ticker operations.
type EntitiesRepository struct {
	db *supabase.Client
}

// NewEntitiesRepository creates a new EntitiesRepository.
func NewEntitiesRepository(db *supabase.Client) *EntitiesRepository {
	return &EntitiesRepository{db: db}
}

// FetchVideoIDsInWindow fetches video_id values within a published_at window.
func (r *EntitiesRepository) FetchVideoIDsInWindow(startISO, endISO string, limit int) ([]string, error) {
	var rows []map[string]interface{}
	err := r.db.From("videos").
		Select("video_id").
		Gte("published_at", startISO).
		Lte("published_at", endISO).
		Order("published_at", true).
		Limit(limit).
		Execute(&rows)
	if err != nil {
		return nil, fmt.Errorf("fetch video ids in window: %w", err)
	}
	ids := make([]string, 0, len(rows))
	for _, row := range rows {
		if id := supabase.GetStr(row, "video_id"); id != "" {
			ids = append(ids, id)
		}
	}
	return ids, nil
}

// FetchSummariesForVideoIDs fetches ticker summaries for given video IDs.
func (r *EntitiesRepository) FetchSummariesForVideoIDs(videoIDs []string, limit int) ([]map[string]interface{}, error) {
	if len(videoIDs) == 0 {
		return nil, nil
	}
	var rows []map[string]interface{}
	err := r.db.From("summaries").
		Select("video_id,ticker,summary").
		In("video_id", videoIDs).
		Limit(limit).
		Execute(&rows)
	if err != nil {
		return nil, fmt.Errorf("fetch summaries for video ids: %w", err)
	}
	return rows, nil
}

// FetchRecentVideos fetches video metadata within a published_at window.
func (r *EntitiesRepository) FetchRecentVideos(startISO, endISO string, limit int) ([]map[string]interface{}, error) {
	var rows []map[string]interface{}
	err := r.db.From("videos").
		Select("video_id,published_at,video_url,channel,title").
		Gte("published_at", startISO).
		Lte("published_at", endISO).
		Order("published_at", true).
		Limit(limit).
		Execute(&rows)
	if err != nil {
		return nil, fmt.Errorf("fetch recent videos: %w", err)
	}
	return rows, nil
}

// FetchEntitySummaryRows fetches summaries for a symbol within allowed video IDs, with embedded video metadata.
func (r *EntitiesRepository) FetchEntitySummaryRows(symbol string, allowedVideoIDs []string, limit int) ([]map[string]interface{}, error) {
	if len(allowedVideoIDs) == 0 {
		return nil, nil
	}
	var rows []map[string]interface{}
	err := r.db.From("summaries").
		Select("video_id,ticker,summary,created_at,videos(video_url,video_id,channel,title,published_at)").
		Eq("ticker", symbol).
		Order("created_at", true).
		Limit(limit).
		In("video_id", allowedVideoIDs).
		Execute(&rows)
	if err != nil {
		return nil, fmt.Errorf("fetch entity summary rows: %w", err)
	}
	return rows, nil
}
